### DISCUSSION TITLE
MetricType中COSINE和任何一个IndexType都不匹配的问题  如下图

### DISCUSSION THREAD

**Discussion by LLLiHaotian** [milvus-io/milvus#30988] [Q&A and General discussion] [answered]
我所要插入的向量是768维的float，因此按照milvus官网参数提示（https://milvus.io/api-reference/java/v2.3.x/Misc/IndexType.md），只有图中圈出的几个符合条件，但均与COSINE不匹配
![image](https://github.com/milvus-io/milvus/assets/150894804/89a5d9fe-2e06-4bf8-9c62-4115743c33be)

![image](https://github.com/milvus-io/milvus/assets/150894804/af6edd91-69e8-42fa-bb9e-33c3d8f7ac0e)
![image](https://github.com/milvus-io/milvus/assets/150894804/c3559286-088e-4ac2-965e-83a0f616eb49)
![image](https://github.com/milvus-io/milvus/assets/150894804/3b59a905-6b27-4a41-a54b-964a94dafb63)
![image](https://github.com/milvus-io/milvus/assets/150894804/12f40cf8-0784-42d3-8741-9afe5ed861e3)
![image](https://github.com/milvus-io/milvus/assets/150894804/8e5c45a9-8dbd-413d-99e6-61a330eadd11)
![image](https://github.com/milvus-io/milvus/assets/150894804/c6ef47f4-9dbe-482e-bf91-1eae2bddcaed)
![image](https://github.com/milvus-io/milvus/assets/150894804/04bf0cb7-a196-4a72-b3b2-9b7a7057642f)

**Comment by yhmo** [comment]
COSINE 是2.3.0版本才开始引入： https://milvus.io/docs/release_notes.md#Cosine-Metrics
2.3.0版本里，ANNOY的索引已经不再支持，文档里也已去掉。而你看的文档里还有ANNOY，说明你用的可能不是2.3.x版本。
我试了一下2.3.10版本的COSINE，所有float的index type都支持，没有报错，功能正常。

**Reply by LLLiHaotian** [reply]
您好，我尝试了python的3.6、3.7、3.8、3.10的版本，connections.connect("default", host="localhost", port="19530")在连接时候都不能正常连接，依然报错RPC error: [__internal_register], <MilvusException: (code=1, message=this version of sdk is incompatible with server, please downgrade your sdk or upgrade your server)>, <Time:{'RPC start': '2024-03-05 11:13:00.669731', 'RPC error': '2024-03-05 11:13:00.671213'}>
请问您的python版本是什么呢？
期待您的回答

**Comment by LLLiHaotian** [comment]
感谢您的回答。
我最开始安装的是pymilvus==2.3.0，但由于在connections.connect("default", host="localhost", port="19530")连接数据库的过程中给报错MilvusException: <MilvusException: (code=1, message=this version of sdk is incompatible with server, please downgrade your sdk or upgrade your server)>，于是我才降低版本到pymilvus==2.2.8。根据报错显示sdk的版本不匹配，请问不同版本的milvus对应的sdk版本是什么？
另外，我在2.28版本下执行创建索引操作出现hang，仅10个768维的向量进行创建索引就出现程序挂起不动的情况，请问这又是什么原因呢？
index = {
    "index_type": "IVF_FLAT",
    "metric_type": "L2",
    "params": {"nlist": 128},
}
# 构建索引
hello_milvus.create_index("embeddings", index)


期待您的再次回答！谢谢！

**Reply by yhmo** [reply]
pymilvus的项目页面里有把版本匹配表：https://github.com/milvus-io/pymilvus?tab=readme-ov-file#compatibility
文档里也有版本匹配表： https://milvus.io/api-reference/pymilvus/v2.3.x/About.md
通常，2.2.x的sdk配2.2.x的server，2.3.x的sdk配2.3.x的server。

不匹配的sdk和server版本，很有可能在调用接口的时候报错或者hang，因为传输协议有可能有变化。

**Reply by LLLiHaotian** [reply]
您好，方便问一下您使用的python版本是多少吗

**Reply by yhmo** [reply, marked as answer]
Milvus: v2.3.10
pymilvus: v2.3.6


连接之后check一下服务端的版本号，也许你以为server是2.3，但实际上可能是你之前跑了个2.2没关闭
connections.connect(host='localhost', port='19530')
print(utility.get_server_version())

**Reply by LLLiHaotian** [reply]
的确，我以为安装的最新版本，已经重新安装了2.3.10，pymilvus是2.3.6，但是现在连接仍然失败，（按照csdn的思路输入命令export http_proxy=""关闭了网络代理），仍然报错如下MilvusException: <MilvusException: (code=2, message=Fail connecting to server on localhost:19530. Timeout)>
请问这个问题之前遇到过吗
![image](https://github.com/milvus-io/milvus/assets/150894804/8c1f956c-2035-454f-82c0-075ef627bfd6)
![image](https://github.com/milvus-io/milvus/assets/150894804/8453b387-b9ae-448b-b319-962b85cef89e)

**Reply by yhmo** [reply]
先确定所有milvus的进程都关闭
用 docker ps -a 命令查看所有容器，把跟milvus的容器全部干掉，用 docker rm
用 ps -e | grep milvus 命令看有没有milvus 进程在运行，有的话就 kill -9 干掉
找一个空的目录，wget https://github.com/milvus-io/milvus/releases/download/v2.3.10/milvus-standalone-docker-compose.yml -O docker-compose.yml 然后 docker-compose up -d 启动

然后再尝试连接

如果这样还不行的话我也帮不了你了，换台电脑吧

**Reply by LLLiHaotian** [reply]
干掉所有之后
![image](https://github.com/milvus-io/milvus/assets/150894804/03b2ab81-f890-4b51-a781-fdc4c386f65f)

不知道docker-compose up -d 启动后这有个error有没有影响
![image](https://github.com/milvus-io/milvus/assets/150894804/88deec97-bca1-42e5-9d86-959fe9bad39c)

**Reply by yhmo** [reply]
minio的端口9001被占用，也许是有旧的minio容器存在
docker ps -a 看是否有多个minio容器，关掉不在运行的那个
然后docker-compose restart重启milvus服务

**Reply by yhmo** [reply]
如果不行就把电脑重启

**Reply by LLLiHaotian** [reply]
![image](https://github.com/milvus-io/milvus/assets/150894804/5c43eaec-4e33-4251-9da3-c763cb6efcf0)
![image](https://github.com/milvus-io/milvus/assets/150894804/2e8a9b13-74cd-4267-b221-54ac69f40bea)

理论上应该是没有duo多个minio容器了

**Reply by yhmo** [reply]
如果还连接不上，用 docker logs [container id] 获取milvus容器的日志，然后看有什么error

**Comment by LLLiHaotian** [comment]
感谢您的回答！
关于sdk的版本希望您能提供一下具体的查看和改变的方法

另外，以上提到的创建索引程序被挂起不停的问题请问是否有相应的解决办法？是不是在新版本中就不会这样了？

**Reply by yhmo** [reply]
python中查看某个库的版本，使用pip3 show XXX 命令

 ```
➜  tls git:(ma) ✗ pip3 show pymilvus   
Name: pymilvus
Version: 2.3.6
Summary: Python Sdk for Milvus
Home-page: 
Author: 
Author-email: Milvus Team <milvus-team@zilliz.com>
License: 
Location: /home/yhmo/git-go/pymilvus/venv10/lib/python3.10/site-packages
Requires: environs, grpcio, minio, pandas, protobuf, pyarrow, requests, setuptools, ujson
Required-by: 

```

安装指定版本  pip3  install pymilvus==2.3.5
卸载 pip3 uninstall pymilvus

这些都是python下的基本操作。

创建索引是个常规操作，正常的版本匹配并不会出问题。如果版本匹配是对的，那就是服务端有问题，或者是安装问题，或者是第三方的服务(etcd, pulsar,minio)不正常，需要查看服务端的日志。
对于使用docker-compose安装的standalone，使用 docker logs [milvus_container_id] 来输出milvus server的日志，查看milvus容器的id可使用docker ps 命令，这些是docker的基本操作。

**Comment by LLLiHaotian** [comment]
感谢您的回答。
可能我的表达有误，我明白查看pymilvus的版本的基本python操作，我的版本与你提供的一致，但仍在连接时报错
RPC error: [__internal_register], <MilvusException: (code=1, message=this version of sdk is incompatible with server, please downgrade your sdk or upgrade your server)>, <Time:{'RPC start': '2024-03-04 13:49:35.390283', 'RPC error': '2024-03-04 13:49:35.391715'}>
可能时服务端的问题

**Reply by xiaofan-luan** [reply]
You will need to upgrade your server version since your client version is higher than server version

**Reply by LLLiHaotian** [reply]
感谢
